File: backend/src/svap/rag.py
"""
RAG (Retrieval-Augmented Generation) utilities for SVAP pipeline.

Handles document ingestion, chunking, retrieval, and context assembly.
Uses keyword-based retrieval by default. Can be extended with vector
search by configuring an embedding model in config.yaml.
"""

import hashlib
import re
from pathlib import Path

try:
    import tiktoken

    _ENCODER = tiktoken.get_encoding("cl100k_base")

    def count_tokens(text: str) -> int:
        return len(_ENCODER.encode(text))
except ImportError:

    def count_tokens(text: str) -> int:
        return len(text.split()) * 4 // 3  # rough approximation


from svap.storage import SVAPStorage


class DocumentIngester:
    """Ingest documents into the storage layer, chunked for retrieval."""

    def __init__(self, storage: SVAPStorage, config: dict):
        self.storage = storage
        self.chunk_size = config.get("rag", {}).get("chunk_size", 1500)
        self.chunk_overlap = config.get("rag", {}).get("chunk_overlap", 200)

    def ingest_file(self, filepath: str, doc_type: str = "other", metadata: dict | None = None):
        """Read a file, store it, and chunk it for retrieval."""
        path = Path(filepath)
        text = path.read_text(encoding="utf-8", errors="replace")
        doc_id = hashlib.sha256(f"{path.name}:{text[:200]}".encode()).hexdigest()[:16]

        self.storage.insert_document(
            doc_id=doc_id,
            filename=path.name,
            doc_type=doc_type,
            full_text=text,
            metadata=metadata,
        )

        chunks = self._chunk_text(text)
        for i, chunk_text in enumerate(chunks):
            chunk_id = f"{doc_id}_c{i:04d}"
            self.storage.insert_chunk(
                chunk_id=chunk_id,
                doc_id=doc_id,
                chunk_index=i,
                text=chunk_text,
                token_count=count_tokens(chunk_text),
            )

        return doc_id, len(chunks)

    def ingest_text(
        self, text: str, filename: str, doc_type: str = "other", metadata: dict | None = None
    ):
        """Ingest raw text directly (no file)."""
        doc_id = hashlib.sha256(f"{filename}:{text[:200]}".encode()).hexdigest()[:16]
        self.storage.insert_document(doc_id, filename, doc_type, text, metadata)

        chunks = self._chunk_text(text)
        for i, chunk_text in enumerate(chunks):
            chunk_id = f"{doc_id}_c{i:04d}"
            self.storage.insert_chunk(chunk_id, doc_id, i, chunk_text, count_tokens(chunk_text))

        return doc_id, len(chunks)

    def ingest_directory(self, dirpath: str, doc_type: str = "other"):
        """Ingest all .txt, .md, and .json files from a directory."""
        results = []
        for ext in ("*.txt", "*.md", "*.json"):
            for path in Path(dirpath).glob(ext):
                doc_id, n_chunks = self.ingest_file(str(path), doc_type)
                results.append({"file": path.name, "doc_id": doc_id, "chunks": n_chunks})
        return results

    def _chunk_text(self, text: str) -> list[str]:
        """Split text into overlapping chunks by token count."""
        # Split on paragraph boundaries first
        paragraphs = re.split(r"\n\s*\n", text)
        chunks = []
        current_chunk = ""
        current_tokens = 0

        for para in paragraphs:
            para_tokens = count_tokens(para)
            if current_tokens + para_tokens > self.chunk_size and current_chunk:
                chunks.append(current_chunk.strip())
                # Overlap: keep the last portion of the current chunk
                overlap_text = self._get_overlap(current_chunk)
                current_chunk = overlap_text + "\n\n" + para
                current_tokens = count_tokens(current_chunk)
            else:
                current_chunk += "\n\n" + para if current_chunk else para
                current_tokens += para_tokens

        if current_chunk.strip():
            chunks.append(current_chunk.strip())

        return chunks if chunks else [text]

    def _get_overlap(self, text: str) -> str:
        """Extract the last ~chunk_overlap tokens of text."""
        words = text.split()
        overlap_words = max(1, self.chunk_overlap * 3 // 4)  # rough token-to-word
        return " ".join(words[-overlap_words:]) if len(words) > overlap_words else text


class ContextAssembler:
    """Assemble context from retrieved chunks for prompt injection."""

    def __init__(self, storage: SVAPStorage, config: dict):
        self.storage = storage
        self.max_chunks = config.get("rag", {}).get("max_context_chunks", 10)

    def retrieve(
        self, query: str, doc_type: str | None = None, max_chunks: int | None = None
    ) -> str:
        """Retrieve relevant chunks and format as context block."""
        limit = max_chunks or self.max_chunks
        chunks = self.storage.search_chunks(query, doc_type=doc_type, limit=limit)

        if not chunks:
            return ""

        context_parts = []
        for chunk in chunks:
            source = chunk.get("filename", "unknown")
            context_parts.append(f"[Source: {source}]\n{chunk['text']}")

        return "\n\n---\n\n".join(context_parts)

    def retrieve_all_of_type(self, doc_type: str) -> str:
        """Return all documents of a given type concatenated (for small corpora)."""
        docs = self.storage.get_all_documents(doc_type)
        return "\n\n===\n\n".join(f"[{d['filename']}]\n{d['full_text']}" for d in docs)

    def format_cases_context(self, cases: list[dict]) -> str:
        """Format case data as structured context for prompts."""
        parts = []
        for c in cases:
            parts.append(
                f"CASE: {c['case_name']}\n"
                f"  Scheme: {c['scheme_mechanics']}\n"
                f"  Exploited Policy: {c['exploited_policy']}\n"
                f"  Enabling Condition: {c['enabling_condition']}\n"
                f"  Scale: ${c.get('scale_dollars', 'unknown')}\n"
                f"  Detection: {c.get('detection_method', 'unknown')}"
            )
        return "\n\n".join(parts)

    def format_taxonomy_context(self, taxonomy: list[dict]) -> str:
        """Format taxonomy as structured context for prompts."""
        parts = []
        for q in taxonomy:
            parts.append(
                f"{q['quality_id']} — {q['name']}\n"
                f"  Definition: {q['definition']}\n"
                f"  Recognition Test: {q['recognition_test']}\n"
                f"  Exploitation Logic: {q['exploitation_logic']}"
            )
        return "\n\n".join(parts)

File: backend/shared/src/rag.rs
//! RAG (Retrieval-Augmented Generation) utilities.
//!
//! Handles document ingestion, chunking, retrieval, and context assembly.
//! Uses keyword-based retrieval by default (matching the Python implementation).

use regex::Regex;
use sha2::{Digest, Sha256};
use tokio_postgres::Client;

use crate::db;
use crate::types::{Case, Config, TaxonomyQuality};

/// Count tokens using tiktoken cl100k_base encoding.
pub fn count_tokens(text: &str) -> usize {
    tiktoken_rs::cl100k_base()
        .map(|bpe| bpe.encode_with_special_tokens(text).len())
        .unwrap_or_else(|_| text.split_whitespace().count() * 4 / 3)
}

// ── Document Ingester ────────────────────────────────────────────────────

pub struct DocumentIngester {
    chunk_size: usize,
    chunk_overlap: usize,
}

impl DocumentIngester {
    pub fn new(config: &Config) -> Self {
        Self {
            chunk_size: config.rag.chunk_size,
            chunk_overlap: config.rag.chunk_overlap,
        }
    }

    /// Ingest raw text directly. Returns (doc_id, n_chunks).
    pub async fn ingest_text(
        &self,
        client: &Client,
        text: &str,
        filename: &str,
        doc_type: &str,
        metadata: Option<&serde_json::Value>,
    ) -> Result<(String, usize), Box<dyn std::error::Error + Send + Sync>> {
        let mut hasher = Sha256::new();
        hasher.update(format!("{}:{}", filename, &text[..text.len().min(200)]));
        let doc_id = format!("{:x}", hasher.finalize())[..16].to_string();

        db::insert_document(client, &doc_id, filename, doc_type, text, metadata).await?;

        let chunks = self.chunk_text(text);
        for (i, chunk_text) in chunks.iter().enumerate() {
            let chunk_id = format!("{}_c{:04}", doc_id, i);
            let token_count = count_tokens(chunk_text) as i32;
            db::insert_chunk(
                client,
                &chunk_id,
                &doc_id,
                i as i32,
                chunk_text,
                token_count,
            )
            .await?;
        }

        Ok((doc_id, chunks.len()))
    }

    fn chunk_text(&self, text: &str) -> Vec<String> {
        let paragraphs: Vec<&str> = Regex::new(r"\n\s*\n").unwrap().split(text).collect();
        let mut chunks = Vec::new();
        let mut current_chunk = String::new();
        let mut current_tokens = 0;

        for para in &paragraphs {
            let para_tokens = count_tokens(para);
            if current_tokens + para_tokens > self.chunk_size && !current_chunk.is_empty() {
                chunks.push(current_chunk.trim().to_string());
                let overlap_text = self.get_overlap(&current_chunk);
                current_chunk = format!("{}\n\n{}", overlap_text, para);
                current_tokens = count_tokens(&current_chunk);
            } else {
                if current_chunk.is_empty() {
                    current_chunk = para.to_string();
                } else {
                    current_chunk.push_str("\n\n");
                    current_chunk.push_str(para);
                }
                current_tokens += para_tokens;
            }
        }

        if !current_chunk.trim().is_empty() {
            chunks.push(current_chunk.trim().to_string());
        }

        if chunks.is_empty() {
            vec![text.to_string()]
        } else {
            chunks
        }
    }

    fn get_overlap(&self, text: &str) -> String {
        let words: Vec<&str> = text.split_whitespace().collect();
        let overlap_words = (self.chunk_overlap * 3 / 4).max(1);
        if words.len() > overlap_words {
            words[words.len() - overlap_words..].join(" ")
        } else {
            text.to_string()
        }
    }
}

// ── Context Assembler ────────────────────────────────────────────────────

pub struct ContextAssembler {
    max_chunks: usize,
}

impl ContextAssembler {
    pub fn new(config: &Config) -> Self {
        Self {
            max_chunks: config.rag.max_context_chunks,
        }
    }

    /// Retrieve relevant chunks and format as context block.
    pub async fn retrieve(
        &self,
        client: &Client,
        query: &str,
        doc_type: Option<&str>,
        max_chunks: Option<usize>,
    ) -> Result<String, Box<dyn std::error::Error + Send + Sync>> {
        let limit = max_chunks.unwrap_or(self.max_chunks);
        let chunks = db::search_chunks(client, query, doc_type, limit).await?;

        if chunks.is_empty() {
            return Ok(String::new());
        }

        let context_parts: Vec<String> = chunks
            .iter()
            .map(|chunk| {
                let source = chunk.filename.as_deref().unwrap_or("unknown");
                format!("[Source: {}]\n{}", source, chunk.text)
            })
            .collect();

        Ok(context_parts.join("\n\n---\n\n"))
    }

    /// Return all documents of a given type concatenated.
    pub async fn retrieve_all_of_type(
        &self,
        client: &Client,
        doc_type: &str,
    ) -> Result<String, Box<dyn std::error::Error + Send + Sync>> {
        let docs = db::get_all_documents(client, Some(doc_type)).await?;
        let parts: Vec<String> = docs
            .iter()
            .map(|d| {
                format!(
                    "[{}]\n{}",
                    d.filename.as_deref().unwrap_or("unknown"),
                    d.full_text
                )
            })
            .collect();
        Ok(parts.join("\n\n===\n\n"))
    }

    /// Format case data as structured context for prompts.
    pub fn format_cases_context(cases: &[Case]) -> String {
        cases
            .iter()
            .map(|c| {
                format!(
                    "CASE: {}\n  Scheme: {}\n  Exploited Policy: {}\n  Enabling Condition: {}\n  Scale: ${}\n  Detection: {}",
                    c.case_name,
                    c.scheme_mechanics,
                    c.exploited_policy,
                    c.enabling_condition,
                    c.scale_dollars.map(|v| format!("{:.0}", v)).unwrap_or_else(|| "unknown".to_string()),
                    c.detection_method.as_deref().unwrap_or("unknown"),
                )
            })
            .collect::<Vec<_>>()
            .join("\n\n")
    }

    /// Format taxonomy as structured context for prompts.
    pub fn format_taxonomy_context(taxonomy: &[TaxonomyQuality]) -> String {
        taxonomy
            .iter()
            .map(|q| {
                format!(
                    "{} -- {}\n  Definition: {}\n  Recognition Test: {}\n  Exploitation Logic: {}",
                    q.quality_id, q.name, q.definition, q.recognition_test, q.exploitation_logic,
                )
            })
            .collect::<Vec<_>>()
            .join("\n\n")
    }
}
